# Creating connection to MySQL with Python using MySql connector

In [105]:
import mysql.connector as connection

In [106]:
mydb = connection.connect(host="localhost",user="root", passwd="Iamwhatiam21@",use_pure=True)

In [107]:
mydb

In [122]:
cursor = mydb.cursor()

In [123]:
cursor.execute("use naveendb1")

# Importing the provided movies.csv file into the movies_new4 table in Database

In [6]:
import pandas as pd
import numpy as np

In [7]:
df=pd.read_csv(r'C:\Users\91746\Desktop\Machine Learning Projects\sql + python\Movies SQL\movies.csv')

In [8]:
df

11                                          Star Wars  76.496
0         62                              2001: A Space Odyssey  39.064
1        152                      Star Trek: The Motion Picture  26.551
2        154                    Star Trek II: The Wrath of Khan  17.266
3        157                Star Trek III: The Search for Spock  17.008
4        168                      Star Trek IV: The Voyage Home  17.747
...      ...                                                ...     ...
2478  724435  Star Wars: The Clone Wars - The Siege of Manda...   1.792
2479  724825      AGFA MYSTERY MIXTAPE #4: FOLLOW YOUR OWN STAR   2.744
2480  724968                         Bad Stars: Chemical Valley   0.600
2481  725353                                       Jasmine Star   0.600
2482  726152                                       The Ape Star   1.925

[2483 rows x 3 columns]

In [9]:
df.columns= ["id","movies","score"]

In [73]:
df.head()

id                               movies   score
0   62                2001: A Space Odyssey  39.064
1  152        Star Trek: The Motion Picture  26.551
2  154      Star Trek II: The Wrath of Khan  17.266
3  157  Star Trek III: The Search for Spock  17.008
4  168        Star Trek IV: The Voyage Home  17.747

In [11]:
cursor = mydb.cursor()

In [11]:
cursor.execute("create table movies_new5 (id int,movie varchar(1000),score real)")

In [12]:
for i,row in df.iterrows():
    query= " insert into movies_new5 values(%s,%s,%s)"
    cursor.execute(query, tuple(row))

In [96]:
cursor.execute("select * from movies_new5")

In [97]:
cursor.fetchall()

[(62, '2001: A Space Odyssey', 39.064),
 (152, 'Star Trek: The Motion Picture', 26.551),
 (154, 'Star Trek II: The Wrath of Khan', 17.266),
 (157, 'Star Trek III: The Search for Spock', 17.008),
 (168, 'Star Trek IV: The Voyage Home', 17.747),
 (172, 'Star Trek V: The Final Frontier', 18.588),
 (174, 'Star Trek VI: The Undiscovered Country', 10.975),
 (193, 'Star Trek: Generations', 10.921),
 (199, 'Star Trek: First Contact', 22.78),
 (200, 'Star Trek: Insurrection', 17.369),
 (201, 'Star Trek: Nemesis', 22.021),
 (212, 'Arsenic and Old Lace', 12.55),
 (411,
  'The Chronicles of Narnia: The Lion, the Witch and the Wardrobe',
  31.61),
 (563, 'Starship Troopers', 22.141),
 (679, 'Aliens', 29.666),
 (935,
  'Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb',
  21.628),
 (940, 'The Lady Vanishes', 14.034),
 (1410, 'Dark Star', 8.462),
 (1771, 'Captain America: The First Avenger', 25.99),
 (1884, 'The Ewok Adventure', 11.998),
 (1891, 'The Empire Strikes Back', 27.13),


# Importing the provided movies_cast.csv file into the movies_cast table in Database

In [15]:
# q1_b
# 2. movie_cast
# --1. movie_id (integer)
# --2. cast_id (integer)
# --3. cast_name (text)
# --4. birthday (text)
# --5. popularity (real)

In [16]:
cursor.execute("create table movie_cast1 (movie_id int,cast_id int,cast_name varchar(1000),birthday varchar(1000),popularity real)")

In [17]:
df2=pd.read_csv(r'C:\Users\91746\Desktop\Machine Learning Projects\sql + python\Movies SQL\movie_cast.csv')

In [18]:
df2.columns=["movie_id","cast_id","cast_name","birthday","popularity"]

In [84]:
df2

movie_id  cast_id          cast_name  birthday  popularity
0            11        3      Harrison Ford   7/13/42      14.512
1            11        4      Carrie Fisher  10/21/56       8.905
2            11        5      Peter Cushing   5/26/13       6.350
3            11        6    Anthony Daniels   2/21/46       7.083
4            11      130        Kenny Baker   8/24/34       3.109
...         ...      ...                ...       ...         ...
23374    726152   231318   Melinda Kinnaman   11/9/71       1.495
23375    726152   576510      Gunnar Bohlin      None       0.600
23376    726152  2717973  Rebecca Gerstmann      None       0.600
23377    726152  2717974   Jack Bergenholtz      None       0.600
23378    726152  2717976         Per Broman      None       0.600

[23379 rows x 5 columns]

In [112]:
for i,rows in df2.iterrows():
    cursor.execute("insert into movie_cast1 value(%s,%s,%s,%s,%s)",tuple(rows))
    

In [117]:
cursor.execute("select * from movie_cast1")

In [118]:
cursor.fetchall()

[(11, 3, 'Harrison Ford', '7/13/42', 14.512),
 (11, 4, 'Carrie Fisher', '10/21/56', 8.905),
 (11, 5, 'Peter Cushing', '5/26/13', 6.35),
 (11, 6, 'Anthony Daniels', '2/21/46', 7.082999999999999),
 (11, 130, 'Kenny Baker', '8/24/34', 3.109),
 (11, 663, 'William Hootkins', '7/5/48', 1.591),
 (11, 3044, 'Scott Beach', '1/13/31', 1.788),
 (11, 4945, 'Joe Johnston', '5/13/50', 3.074),
 (11, 7727, 'Phil Tippett', 'None', 1.62),
 (11, 10657, 'Shane Rimmer', '5/28/29', 2.6660000000000004),
 (11, 12248, 'Alec Guinness', '4/2/14', 5.599),
 (11, 12829, 'Richard LeParmentier', '7/16/46', 0.98),
 (11, 15152, 'James Earl Jones', '1/17/31', 6.978),
 (11, 17356, 'Garrick Hagon', '9/27/39', 2.036),
 (11, 24278, 'Bill Weston', '5/29/41', 1.4),
 (11, 24342, 'David Prowse', '7/1/35', 1.016),
 (11, 24343, 'Peter Mayhew', '5/19/44', 3.2539999999999996),
 (11, 33032, 'Phil Brown', '4/30/16', 1.686),
 (11, 42570, 'Don Henderson', '11/10/32', 1.006),
 (11, 47401, 'Drewe Henley', 'None', 1.128),
 (11, 47698, 'De

# Calculating a proportion. Finding the proportion of movies having a score > 10 and that has ‘war’ in
# the name

In [23]:
query = "select count(*)*100/(select count(*) from movies_new5)  from movies_new5 where score > 10 and movie regexp 'war?'"

In [24]:
cursor.execute(query)

In [25]:
cursor.fetchall()

[(Decimal('0.5638'),)]

# only 0.5638% of total movies satisfies above query

# Finding the most prolific actors. Listing 5 cast members with the highest number of movie
# appearances that have a popularity > 10. Sorting the results by the number of appearances in descending
# order, then by cast_name in alphabetical order.

In [28]:
query="select cast_name,count(cast_name) from movie_cast1 where popularity > 10 group by cast_name order by count(cast_name) desc limit 5 "
cursor.execute(query)

In [29]:
cursor.fetchall()

[('Mark Hamill', 16),
 ('Harrison Ford', 14),
 ('Billy Dee Williams', 12),
 ('Patrick Stewart', 11),
 ('Samuel L. Jackson', 9)]

# Finding the highest scoring movies with the smallest cast. Listing the 5 highest-scoring movies that
# have the fewest cast members. Sorting the results by score in descending order, then by number of cast
# members in ascending order, then by movie name in alphabetical order. 

In [35]:
query=("select movie,score,count(cast_name) from movies_new5 inner join movie_cast1 on movies_new5.id = movie_cast1.movie_id group by movie order by score desc, count(cast_name) desc,movie asc limit 10")

In [36]:
cursor.execute(query)

# output:

In [37]:
cursor.fetchall()

[('Interstellar', 99.39200000000001, 37),
 ('Star Wars: The Rise of Skywalker', 69.039, 112),
 ('Star Wars: The Last Jedi', 42.653, 115),
 ('2001: A Space Odyssey', 39.064, 60),
 ('Solo: A Star Wars Story', 37.443000000000005, 111),
 ('Star Wars: The Force Awakens', 36.543, 191),
 ('The Chronicles of Narnia: The Lion, the Witch and the Wardrobe', 31.61, 62),
 ('Aliens', 29.666, 27),
 ('Bolt', 28.743000000000002, 46),
 ('Star Wars: Episode I - The Phantom Menace', 28.646, 121)]

# Getting highest scoring actors. Finding the top ten cast members who have the highest average movie
# scores. Formatting all decimals to two place.
# ▪ Sorting the output by average score in descending order, then by cast_name in alphabetical order.
# ▪ not including the movies with score <25 in the average score calculation.
# ▪ Excluding cast members who have appeared in two or fewer movies.
#  Output format and sample values (cast_id,cast_name,average_score):
# 8822,Julia Roberts,53.00

In [132]:
query="select cast_id, cast_name,format(avg(score),2) from movies_new5 inner join movie_cast1 on movies_new5.id=movie_cast1.movie_id group by cast_name having avg(score) > 25 AND count(cast_id) >2 order by avg(score) desc,cast_name asc limit 10"
cursor.execute(query)

# Output:

In [133]:
cursor.fetchall()

[(1329041, 'Brian Herring', '46.42'),
 (1234529, 'Dave Chapman', '46.42'),
 (1709041, 'Joonas Suotamo', '46.42'),
 (100085, 'Paul Kasey', '43.90'),
 (1729666, 'Bill Kipsang Rotich', '43.30'),
 (1249957, 'Mike Quinn', '43.14'),
 (1333, 'Andy Serkis', '42.95'),
 (47698, 'Denis Lawson', '40.16'),
 (181879, 'Gerald W. Abrams', '39.59'),
 (17052, 'Topher Grace', '39.22')]